### PR TITLE
Support redirect flows

### PR DESCRIPTION
**Pull request by stavros-wb**
• Builds on https://github.com/bergie/passport-saml/pull/16
  • Reverts 638ce6e4e257d99bf1f5b59c2b76f0890744ded9
  • Implements https://github.com/bergie/passport-saml/issues/191

### CODE REVIEW

File: lib/passport-saml/saml.js [hunk 1]
```diff
@@ -689,6 +680,150 @@ SAML.prototype.validatePostResponse = function (container, callback) {
   .done();
 };
 
+SAML.prototype.validateInResponseTo = function(inResponseTo) {
+  if(this.options.validateInResponseTo){
```
Comment (SokratisVidros) [inline lib/passport-saml/saml.js]: nit: missing spaces around `()`

File: lib/passport-saml/saml.js [hunk 2]
```diff
@@ -689,6 +680,150 @@ SAML.prototype.validatePostResponse = function (container, callback) {
   .done();
 };
 
+SAML.prototype.validateInResponseTo = function(inResponseTo) {
```
Comment (SokratisVidros) [inline lib/passport-saml/saml.js]: nit: missing space after `function`

File: lib/passport-saml/saml.js [hunk 3]
```diff
@@ -689,6 +680,150 @@ SAML.prototype.validatePostResponse = function (container, callback) {
...
+
+                var hasValidQuerySignature = certs.some(function (cert) {
+                  return validateSignatureForRedirect(
+                    urlString, signature, container.SigAlg, cert
+                  );
+                });
+
+                if (!hasValidQuerySignature) {
+                  throw 'Invalid signature';
+                }
```
Comment (SokratisVidros) [inline lib/passport-saml/saml.js]: Should we return a `Q()` here as well to have a common method return value?

File: lib/passport-saml/strategy.js [hunk 4]
```diff
@@ -78,10 +78,12 @@ Strategy.prototype.authenticate = function (req, options) {
     }
   }
 
-  if (req.body && req.body.SAMLResponse) {
-      this._saml.validatePostResponse(req.body, validateCallback);
+  if (req.query && (req.query.SAMLResponse || req.query.SAMLRequest)) {
```
Comment (SokratisVidros) [inline lib/passport-saml/strategy.js:81]: Awesome! 🍦

File: test/tests.js [hunk 5]
```diff
@@ -1667,4 +1673,138 @@ describe( 'passport-saml /', function() {
...
+      it('errors if bad xml', function(done) {
+        var body = {
+          SAMLRequest: "asdf"
+        };
+        samlObj.validateRedirect(body, function(err) {
+          should.exist(err);
+          done();
+        });
+      });
+      it('errors if samlIssuer is set and wrong issuer', function(done) {
```
Comment (SokratisVidros) [inline test/tests.js]: nit: `...issuer is wrong...`

File: test/tests.js [hunk 6]
```diff
@@ -1667,4 +1673,138 @@ describe( 'passport-saml /', function() {
...
+      });
+      it('errors if request has expired', function(done) {
+        this.clock.restore();
+        samlObj.validateRedirect(this.request, function(err) {
+          should.exist(err);
+          err.message.should.eql('SAML assertion expired');
+          done();
+        });
+      });
+      it('errors if bad signature', function(done) {
```
Comment (SokratisVidros) [inline test/tests.js]: `if request has a bad signature`

File: lib/passport-saml/saml.js [hunk 7]
```diff
@@ -689,6 +680,150 @@ SAML.prototype.validatePostResponse = function (container, callback) {
...
+        callback(err);
+      });
+    });
+  });
+};
+
+function validateSignatureForRedirect (urlString, signature, alg, cert) {
+  var supportedAlgs = crypto.getHashes().filter(
+    function(h) { return new RegExp(h).test(alg) }
+  )
```
Comment (SokratisVidros) [inline lib/passport-saml/saml.js]: nit: missing semicolon

File: lib/passport-saml/saml.js [hunk 8]
```diff
@@ -689,6 +680,150 @@ SAML.prototype.validatePostResponse = function (container, callback) {
...
+      if (err) {
+        return callback(err);
+      }
+
+      Q.fcall(function () {
+        return type === 'SAMLResponse' ? self.verifyLogoutResponse(doc) : self.verifyLogoutRequest(doc);
+      })
+      .then(function() {
+        var signature = container.Signature;
+        if (signature) {
```
Comment (SokratisVidros) [inline lib/passport-saml/saml.js]: How about factoring out this code to a `verifySignature` method, similarly to `verifyIssuer` method? That way the promise chain would look clearer.

Also, what happens if signature is not present?

Moreover, nit: `if (signature && self.options.cert)`

**Review by SokratisVidros — CHANGES_REQUESTED** [review]

File: lib/passport-saml/saml.js (hunk 3 above)
Comment (stavros-wb) [reply lib/passport-saml/saml.js]: I threw an error so it will resolve with fail

**Pull request closed without merging**